[WIP] Conversion of Preferences/preview to mvvm

calixtus opened the pull request:
follow-up to #5033 

This is my first shot on the PreviewTab in the Preferences and it has been a real nightmare. What started as an i-thought-so-easy-sunday-morning-project became a horrorstory of listViews, impossible bindings and codeAreas. codeArea? Yes. Syntax-highlighting is real fancy. But there are questions left.

1. Can I use the syntax-highlighting-stuff of RichTextFX, or to be more precise: The sample, the creators of RichTextFX credited (see the method computeHighlighting )

2. What does PreviewCycle do, what should it do?

3. What should the default-button do?

The store-logic somehow copied from the original PreviewTab

<img width="500" alt="previewtab" src="https://user-images.githubusercontent.com/50491877/59569730-ed727300-908d-11e9-8438-0c09a703827f.png">

----

- [ ] Change in CHANGELOG.md described
- [ ] Tests created for changes
- [x] Manually tested changed features in running JabRef
- [x] Screenshots added in PR description (for bigger UI changes)
- [x] Ensured that [the git commit message is a good one](https://github.com/joelparkerhenderson/git_commit_message)
- [ ] Check documentation status (Issue created for outdated help page at [help.jabref.org](https://github.com/JabRef/help.jabref.org/issues)?)

Siedlerchr commented:
First of all thanks for your efforts.
The preview tab displays JabRef custom html style or the CSL styles that are converted to html, but which you can't modify.

Preview cycle method allows you to cycle through the styles in the entry editor defined in the list on the right. 
I think it's F9.

Siedlerchr commented:
Furthermore the position of the current selected preview Style is stored in n the prefs. This info is used for the preview cycle function. Complex stuff 😈

Siedlerchr commented on src/main/java/org/jabref/gui/preferences/PreviewTabView.java:
```diff
@@ -0,0 +1,250 @@
+
+    /**
+     * This was originally created by Carlos Martins (github: @cemarins )
+     * Can this be used here? (BSD-2clause License)
```
Yep. BSD2-clause aka FreeBSD is compatible

calixtus replied:
So I can cut the comment? Where do I give credit to him?

Siedlerchr replied:
I would suggest extending the javadoc comment, e.g. author copyright, and link to the code 
Not sure if you need the whole Disclaimer or licence text.
https://opensource.org/licenses/BSD-2-Clause
That should be enough. We already link to the library in our external-libraries.txt file

Siedlerchr commented on src/main/java/org/jabref/gui/preferences/PreviewTabViewModel.java:
```diff
@@ -0,0 +1,188 @@
+
+    @Override
+    public void storeSettings() {
+        PreviewPreferences previewPreferences = Globals.prefs.getPreviewPreferences();
```
preferences.getPreviewPreferences() ? as you already have a JabRef prefs object

calixtus replied:
Thanks, it's been late. 😄

tobiasdiez commented:
That tab is indeed very confusing. A few suggestions (depending on how much time you want to invest):
- Move the "Default" button below the source code (because it is not connected to the list) and rename it to "Reset to default"
- Move to the "up/down" buttons directly in the row of the item as e.g here
![image](https://i.stack.imgur.com/HyISw.png). In some dialogs we already provide a similar item-wise functionality if you want to have a look at the code. Alternatively/in addition support [drag and drop to reorder](https://material.io/design/components/lists.html#behavior)
- Make the "Test/Edit source" feature similar to how editing works here on github: have a tab with the rendered preview and one to edit the source.

Siedlerchr commented:
Drag and drop reorder is implemented in the general tab for moving files up and down as far as I remember

calixtus commented:
Thanks for the comments, i will look into it and resurface the next days again.

calixtus commented:
I got the code-editor and the Preview now in the layout, as @tobiasdiez suggested, but there are still many issues. E.g. there is something wrong with the PreviewViewer: I can change the source for the default 'Preview', but it wont use the changed source in the PreviewViewer until I completly reload the class, even in the MainTable-Preview-View. Also the defaul-button now does not work at all, I think these two issues are connected.

<img width="504" alt="previewtab" src="https://user-images.githubusercontent.com/50491877/59772494-2720c500-92ac-11e9-90a8-26b522bb8b36.png">

Next I'm going to look into the drag'n'drop-stuff. Just wanted you to know, I am still on it. But I got little time these days to work on it.

tobiasdiez commented:
It should work if you call PreviewViewer.setLayout with the updated layout.

calixtus commented:
Thanks, that did the trick on the PreviewPane in the Preferences. Code is somehow ugly with typecasting all around, but i did not had a better quick solution.
Still on my ToDo-List:
- Default button - I think it would be nice not just to have the standard-TextBasedPreviewLayout, but others too (maybe let the user create one). so the default button should not be hardcoded. But I think this goes beyond this PR, so i will hardwire it for now.
- Multiple Selection is broken
- Drag'n'drop or per line-move-arrows. I think of it. Creating arrows would imply changing the ChosenListView to a Table or to fill the ListEntries with HBoxes of the entries and the arrows. Let's see,
Suggestions?

Siedlerchr commented:
If you adapt the html preview, it's stored in prefs. And under the hood the preview works similar or is similar to the custom html export styles

Siedlerchr commented:
Multiple selection is normally handled via the selection model property.  Must be explicitly set

calixtus commented:
I implemented basic drag'n'drop between the two lists, fixed the issues around multiple selection, yet there are still some problems.
- the context-menu. This drives me crazy, i really don't know what to do about this:
![preview_context](https://user-images.githubusercontent.com/50491877/60119774-89336b80-977f-11e9-8f97-5af14f23f415.png)
- the size of the PreviewPane is too large, always, i cannot resize it.

About removing the buttons: this would also take away the ability to control this tab by keyboard. Opinions?

I'm thinking of moving things to the viewModel, but honestly, the longer i work on this, i get more and more confused about what has to go to the viewModel and what to the view. Does somebody have a clear view (😁) about this? Thanks

I am still working on sorting in the chosen ListView. How do I get the index, where the user drops the items? I need general drop-action in the listview, so the user can drop an item in the empty space. Can I add an additional drop-action to the CellFactory or does this interfere with the general drop-action?

Siedlerchr commented on src/main/java/org/jabref/gui/preferences/PreviewTabView.java:
```diff
@@ -0,0 +1,328 @@
+    }
+
+    private void update() {
+        PreviewLayout selectedLayout = viewModel.chosenSelectionModelProperty().getValue().getSelectedItem();
```
This should ideally all be in the viewModel. If you are in doubt what belongs where: All kid of complex logic, etc should be in the viewModel. 
The View itself should be only some "dummy" just providing the controls for bindings etc and adding menus etc

calixtus replied:
Thanks, this info is really helping me, as I read some tutorials about the mvvm-pattern which were not clear about this.

Siedlerchr replied:
See also the example in our wiki. The other important thing: The viewModel does not have any access to the controls/fx nodes directly. The idea is that the viewModel can be tested independently of the View without having to mock the graphical interface. That's why we always use that bindings stuff. 
https://github.com/JabRef/jabref/wiki/Code---JavaFX

Siedlerchr commented on src/main/java/org/jabref/gui/preferences/PreviewTabView.java:
```diff
@@ -0,0 +1,328 @@
+        if (dragboard.hasContent(DragAndDropDataFormats.PREVIEWLAYOUT)) {
+            PreviewLayout draggedLayout = viewModel.findLayoutByNameOrNull((String) dragboard.getContent(DragAndDropDataFormats.PREVIEWLAYOUT));
+            if (draggedLayout != null) {
+                if (dragSourceList != null) {
```
If you look into the code from the LinkedFilesEditor you see how to get the position and add the element at the right position.
```java
  ObservableList<LinkedFileViewModel> items = listView.itemsProperty().get();

        if (dragboard.hasContent(DragAndDropDataFormats.LINKED_FILE)) {

            LinkedFile linkedFile = (LinkedFile) dragboard.getContent(DragAndDropDataFormats.LINKED_FILE);
            LinkedFileViewModel transferedItem = null;
            int draggedIdx = 0;
            for (int i = 0; i < items.size(); i++) {
                if (items.get(i).getFile().equals(linkedFile)) {
                    draggedIdx = i;
                    transferedItem = items.get(i);
                    break;
                }
            }
            int thisIdx = items.indexOf(originalItem);
            items.set(draggedIdx, originalItem);
            items.set(thisIdx, transferedItem);
            success = true;
        }
```

Siedlerchr replied:
Let's say you have a list with A,B,C: You move C to A. You now need to swap the index positions of the entries. I had to use this approach cause I put the serializable model class in the clipboard, so I need to find the corresponding viewModel positions in the listView first.

Siedlerchr commented on src/main/java/org/jabref/gui/preferences/PreviewTabView.java:
```diff
@@ -0,0 +1,328 @@
+        if (layout != null) {
+            ClipboardContent content = new ClipboardContent();
+            Dragboard dragboard = sourceListView.startDragAndDrop(TransferMode.MOVE);
+            content.put(DragAndDropDataFormats.PREVIEWLAYOUT, layout.getName());
```
The PreviewLayout class mus implement the serializable interface.
Note that things like FXproperties or other observables can't be serialized, they must be marked as transient. 
See `LinkedFile` for an example.

calixtus replied:
Ok, i tried making the classes and subclasses of CitationStylePreviewLayout and TextBasedPreviewLayout Serializable, in the Abbreviation-Class i had to make the StringProperties transient. Somehow java is still complaining about a non-serializable-object, but I can see none.

java.lang.IllegalArgumentException: Could not serialize the data
	at com.sun.javafx.tk.quantum.QuantumClipboard.putContentToPeer(QuantumClipboard.java:625) ~[jfxrt.jar:?]
	at com.sun.javafx.tk.quantum.QuantumClipboard.flush(QuantumClipboard.java:269) ~[jfxrt.jar:?]
	at com.sun.javafx.tk.quantum.QuantumToolkit.startDrag(QuantumToolkit.java:1274) ~[jfxrt.jar:?]
	at javafx.scene.Scene$DnDGesture.dragDetectedProcessed(Scene.java:2953) ~[jfxrt.jar:?]
	at javafx.scene.Scene$DnDGesture.process(Scene.java:3022) ~[jfxrt.jar:?]
	at javafx.scene.Scene$DnDGesture.access$8200(Scene.java:2909) ~[jfxrt.jar:?]
	at javafx.scene.Scene$MouseHandler.process(Scene.java:3773) ~[jfxrt.jar:?]
	at javafx.scene.Scene$MouseHandler.access$1500(Scene.java:3485) ~[jfxrt.jar:?]
	at javafx.scene.Scene.impl_processMouseEvent(Scene.java:1762) ~[jfxrt.jar:?]
	at javafx.scene.Scene$ScenePeerListener.mouseEvent(Scene.java:2494) ~[jfxrt.jar:?]
	at com.sun.javafx.tk.quantum.GlassViewEventHandler$MouseEventNotification.run(GlassViewEventHandler.java:394) ~[jfxrt.jar:?]
	at com.sun.javafx.tk.quantum.GlassViewEventHandler$MouseEventNotification.run(GlassViewEventHandler.java:295) ~[jfxrt.jar:?]
	at java.security.AccessController.doPrivileged(Native Method) ~[?:1.8.0_212]
	at com.sun.javafx.tk.quantum.GlassViewEventHandler.lambda$handleMouseEvent$358(GlassViewEventHandler.java:432) ~[jfxrt.jar:?]
	at com.sun.javafx.tk.quantum.QuantumToolkit.runWithoutRenderLock(QuantumToolkit.java:410) ~[jfxrt.jar:?]
	at com.sun.javafx.tk.quantum.GlassViewEventHandler.handleMouseEvent(GlassViewEventHandler.java:431) ~[jfxrt.jar:?]
	at com.sun.glass.ui.View.handleMouseEvent(View.java:555) ~[jfxrt.jar:?]
	at com.sun.glass.ui.View.notifyMouse(View.java:937) ~[jfxrt.jar:?]
	at com.sun.glass.ui.win.WinApplication._runLoop(Native Method) ~[jfxrt.jar:?]
	at com.sun.glass.ui.win.WinApplication.lambda$null$152(WinApplication.java:177) ~[jfxrt.jar:?]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_212]
Caused by: java.io.NotSerializableException: sun.nio.cs.UTF_8
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1184) ~[?:1.8.0_212]
	at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1548) ~[?:1.8.0_212]
	at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1509) ~[?:1.8.0_212]
	at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1432) ~[?:1.8.0_212]
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1178) ~[?:1.8.0_212]
	at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1548) ~[?:1.8.0_212]
	at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1509) ~[?:1.8.0_212]
	at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1432) ~[?:1.8.0_212]
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1178) ~[?:1.8.0_212]
	at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1548) ~[?:1.8.0_212]
	at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1509) ~[?:1.8.0_212]
	at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1432) ~[?:1.8.0_212]
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1178) ~[?:1.8.0_212]
	at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:348) ~[?:1.8.0_212]
	at java.util.ArrayList.writeObject(ArrayList.java:766) ~[?:1.8.0_212]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_212]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_212]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_212]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_212]
	at java.io.ObjectStreamClass.invokeWriteObject(ObjectStreamClass.java:1140) ~[?:1.8.0_212]
	at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1496) ~[?:1.8.0_212]
	at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1432) ~[?:1.8.0_212]
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1178) ~[?:1.8.0_212]
	at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1548) ~[?:1.8.0_212]
	at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1509) ~[?:1.8.0_212]
	at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1432) ~[?:1.8.0_212]
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1178) ~[?:1.8.0_212]
	at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1548) ~[?:1.8.0_212]
	at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1509) ~[?:1.8.0_212]
	at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1432) ~[?:1.8.0_212]
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1178) ~[?:1.8.0_212]
	at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:348) ~[?:1.8.0_212]
	at com.sun.javafx.tk.quantum.QuantumClipboard.putContentToPeer(QuantumClipboard.java:621) ~[jfxrt.jar:?]
	... 20 more

Siedlerchr replied:
Best is to debug and look at which property it fails. It's a bit trial and error, remember this from linked fiel

tobiasdiez replied:
Since the drag and drop is only JabRef-wide (i.e you don't need to support droping on different applications), it is easier to use the `CustomLocalDragboard` (which does not need to serialize the objects). Have a look here: https://github.com/JabRef/jabref/search?q=BIBENTRY_LIST_CLASS&unscoped_q=BIBENTRY_LIST_CLASS

Siedlerchr commented on src/main/java/org/jabref/gui/preferences/PreviewTabView.java:
```diff
@@ -0,0 +1,328 @@
+
+        editArea.clear();
+        editArea.setParagraphGraphicFactory(LineNumberFactory.get(editArea));
+        editArea.textProperty().addListener((observable, oldText, newText) -> editArea.setStyleSpans(0, viewModel.computeHighlighting(newText)));
```
Have you checked if the parent window has setResizable(true) set?

calixtus replied:
The parent window would be the preferences-dialog, the immediate parent of the PreviewViewer is a TabPane. I believe this is something about the webView the Preview is internally shown in.

calixtus commented:
Well, this should be somehow ready to review, I think. The two issues remaining are the size of the PreviewViewer (I believe its an issue of the underlying webpane, but proove me wrong) and the slightly displaced icons in the context-menu, for what i have no clue how to fix that. Anyone with some more jfx-css-experience got some ideas?

calixtus commented:
about failed Travis: somehow the test "setStringAsCollectionWithUpdatedContentOverridesString" is failing now. I have no idea why, i did not change anything close to it.

Siedlerchr commented on src/main/java/org/jabref/gui/util/CustomLocalDragboard.java:
```diff
@@ -63,4 +64,24 @@ public void putBibEntries(List<BibEntry> entries) {
+    /**
+     * Get a List of {@link PreviewLayout} from the dragboard
+     *
+     * @return List of BibEntry or empty list if no entries are avaiable
```
Change to PreviewLayout ;)

Siedlerchr commented:
Thanks for the awesome work! 🥇  Looks good to me so far!

calixtus commented:
Just to remind of this PR, it would be great to have some feedback on the two issues mentioned above: the displaced icons in the ContextMenu and the internal size of the webpane. the size of the webpane is not that big thing, if there is no solution, the user just has to scroll a little bit more to the right. Not perfect, but ... meh...
If there is no solution on the displaced menuicons, I can remove the icons completly.
Besides that, I think everything else is ready to review...

Siedlerchr commented:
Could you please add a screensehot on the misplaced menu items?

calixtus commented:
<img width="496" alt="displaced icons" src="https://user-images.githubusercontent.com/50491877/61276883-3ca4e400-a7b1-11e9-94b5-f8a593798451.png">
<img width="499" alt="unresizeable webpane" src="https://user-images.githubusercontent.com/50491877/61276884-3ca4e400-a7b1-11e9-9145-c83a4bbbbacf.png">

calixtus commented:
Ok, I tried to debug the css-issues a little bit, but I could not figure it out, so i removed the icons. Of course, this is not a solution for the problem, but nobody will ever notice. ;-)
On the other hand, I was finally able to reset the size of the previewPane. Turns out, I had to cast it to a PreviewViewer-Class first.
I also made some minor refactoring on the ViewModel: Instead of testing if the value of the SelectionModelProperty is null I made them an instance of NoSelectionModel as the default value.
Should be ready to merge.

calixtus commented:
Could not resist refactoring a little bit more... 😅

Siedlerchr commented:
I am merging then ;)

calixtus commented:
So this was fun again, at least in the beginning and in the end. In the middle it really hurt.
Before I can work on other tabs, I really have to get my "Lizenz"-thesis done. So it could take about a month, I really have to restrain myself here. 😢 But I will continue, I really don't like leaving things half-done.

Feel free to work on the other tabs. I think, before we start reordering everything, we should convert everything first to mvvm, because otherwise it will become a bit confusing, what is already moved to another prefTab etc.

Siedlerchr commented:
Thanks for your great work and your interest! Regarding your thesis, I think you are not alone. You can ask other JabRef devs about how working on JabRef provides a nice distraction from finishing a thesis....  😈
I think the other tabs should not be that complex

koppor commented:
@calixtus We hope you finished your thesis. Feel free to rejoin for [#hacktoberfest](https://benbarth.github.io/hacktoberfest-swag/).
